MaxText/layers/attentions.py

```diff
@@ -528,7 +528,25 @@ def kv_cache_prefill(
     cached_prefill_key_var, cached_prefill_value_var, cached_prefill_segment_id = self._get_prefill_cache(
         batch, heads, kv_head_size, self.quantize_kvcache
     )
-    self._get_ar_cache(batch, heads, kv_head_size, self.quantize_kvcache)  # initialize it now
+    cached_ar_key_var, cached_ar_value_var, _, _ = self._get_ar_cache(batch, heads, kv_head_size, self.quantize_kvcache)  # initialize it now
```

Comment:
This looks good spiritually but why is it it necessary?

Shouldn't we be doing it here:
https://github.com/google/maxtext/blob/d590328e4fd74e128cf2678d51b848c148fdcc3e/MaxText/layers/attentions.py#L460-L473

(Have you confirmed this change fixes the problem? Could there be a bug in the cache initialization code.)

Reply:
yes, this fixes sharding. Seems we need to call with_logical_constraint to enforce the sharding. Same thing as we did for the ar_cache here:
https://github.com/google/maxtext/blob/bdb1b3e1fca59b632ee5f43d6432d9ce8a15ded3/MaxText/layers/attentions.py#L585-L608

Reply:
`[nn.with_logical_partitioning()](https://github.com/google/flax/blob/main/flax/linen/spmd.py#L331)` only adds partitioning metadata to the param, need to call `nn.with_logical_constraint()` to enforce the sharding.